addons/source-python/packages/source-python/core/__init__.py

```diff
@@ -330,4 +349,79 @@ def check_info_output(output):
     if lines[-1].startswith('-'):
         lines.pop()
 
-    return create_checksum(''.join(lines)) != checksum
\ No newline at end of file
+    return create_checksum(''.join(lines)) != checksum
+
+
+def engine_import(skippables=(), skip_privates=True):
+    """Import engine/game specific objects.
+
+    :param tuple skippables:
+        Names to skip when loading engine/game specific files.
+    :param bool skip_privates:
+        Whether or not private names should be skipped.
+    :raise ImportError:
+        If it was not called from global scope.
+    """
+    f = currentframe().f_back
+    if f.f_locals is not f.f_globals:
+        raise ImportError(
+            '"engine_import" must only be called from global scopes.')
+    skippables = ENGINE_IMPORT_SKIPPABLES + skippables
+    caller = getmodule(f)
+    directory, name = Path(caller.__file__).splitpath()
+    for subfolder in (SOURCE_ENGINE, GAME_NAME):
+        directory /= subfolder
+        if not directory.isdir():
+            break
+        path = directory / name
+        if not path.isfile():
+            continue
+        for attr, obj in run_path(path, f.f_globals, caller.__name__).items():
+            if attr in skippables:
+                continue
+            if (attr == '__doc__' and
+                    getattr(caller, '__doc__', None) is not None):
+                continue
+            if attr == '__all__':
+                if hasattr(caller, '__all__'):
+                    obj = tuple(sorted(set(obj + getattr(caller, '__all__'))))
+            elif skip_privates and attr.startswith('_'):
+                continue
+            elif isclass(obj):
```

Comment:
This line raises NotImplementedError on DOD:S and TF2 servers.  I originally thought this was because we don't have any weapon data for those 2 games, but when I tested Garrysmod, it worked fine.

```python
Traceback (most recent call last):

  File "E:\Servers\tf\tf\addons\source-python\packages\source-python\__init__.py", line 93, in load
    setup_sp_command()

  File "E:\Servers\tf\tf\addons\source-python\packages\source-python\__init__.py", line 329, in setup_sp_command
    from core.command import auth, docs, dump, plugin

  File "E:\Servers\tf\tf\addons\source-python\packages\source-python\core\command\__init__.py", line 20, in <module>
    from commands.typed import TypedServerCommand

  File "E:\Servers\tf\tf\addons\source-python\packages\source-python\commands\typed.py", line 30, in <module>
    from filters.players import parse_filter

  File "E:\Servers\tf\tf\addons\source-python\packages\source-python\filters\players.py", line 26, in <module>
    from players.entity import Player

  File "E:\Servers\tf\tf\addons\source-python\packages\source-python\players\entity.py", line 1021, in <module>
    engine_import()

  File "E:\Servers\tf\tf\addons\source-python\packages\source-python\core\__init__.py", line 390, in engine_import
    elif isclass(obj):

  File "E:\Servers\tf\tf\addons\source-python\Python3\inspect.py", line 78, in isclass
    return isinstance(object, type)

  File "E:\Servers\tf\tf\addons\source-python\packages\source-python\weapons\default.py", line 29, in __getattribute__
    'No support for game "{0}"'.format(GAME_NAME))

NotImplementedError: No support for game "tf"
```

If I wrap that isclass in a try/except, it works fine, but I'm not sure if there are any negative consequences of doing so.

Reply:
This happens because `isclass` is internally calling `isinstance` which attempts to get the `__class__` attribute of the object but since `NoWeaponManager` is raising into `__getattribute__`, it fails. The same would occurs with:
```python
from weapons.default import NoWeaponManager
mgr = NoWeaponManager()
mgr.__class__
```
I don't believe we should wrap the exception because it is working as intended. That class is designed to raise in such scenario and I believe it should rather raise into `__getattr__` instead so that it only fails when attempting to get attributes that are not contained in the `__dict__` instance of the class itself.